### PR TITLE
File lock for svn

### PR DESCRIPTION
**Pull request by pfeuffer**
## Proposed changes

Implements the file lock command for SVN.

### Your checklist for this pull request

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

**Contributor**:
- [x] PR is well described and the description can be used as a commit message on squash
- [x] New code is covered with unit tests
- [x] [Changelog entry file](https://github.com/scm-manager/changelog#changelog-entry-files) created in `gradle/changelog` or CHANGELOG.md is updated for plugins
- [x] Feature has been tested with different permissions

**Reviewer**:
- [ ] The clean code principles are respected ([CleanCode](https://clean-code-developer.com/virtues/))
- [ ] All new code/logic is implemented on the right spot / "Should this be done here?"
- [ ] UI changes fits into the layout
- [ ] The UI views / components are responsive (mobile views)
- [ ] Correct translations are available

### Checklist for branch merge request (not required for forks)

- [ ] Branch is green/blue on [Jenkins](https://oss.cloudogu.com/jenkins/)
- [ ] Quality Gate passed on [SonarQube](https://sonarcloud.io/organizations/scm-manager/projects)

### CODE REVIEW

File: scm-plugins/scm-svn-plugin/src/main/java/sonia/scm/repository/spi/SvnFileLockCommand.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+  }
+
+  private void doUnlock(UnlockCommandRequest request, Predicate<SvnFileLock> predicate) throws SVNException {
+    String fileToUnlock = request.getFile();
+    SVNRepository svnRepository = open();
+    initializeAuthentication(svnRepository);
+    Optional<SvnFileLock> fileLock = getFileLock(fileToUnlock, svnRepository);
+    if (fileLock.isPresent()) {
+      SvnFileLock lock = fileLock.get();
+      if (!request.isForce() && !getCurrentUser().equals(lock.getUserId()) || !predicate.test(lock)) {
```
Comment (eheimbuch) [inline scm-plugins/scm-svn-plugin/src/main/java/sonia/scm/repository/spi/SvnFileLockCommand.java:101]: Wrap this check into a new method? `shouldPreventUnlock`?

File: scm-plugins/scm-svn-plugin/src/main/java/sonia/scm/repository/spi/SvnFileLockCommand.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+      }
+      svnRepository.unlock(singletonMap(fileToUnlock, lock.getId()), request.isForce(), null);
+    }
+  }
+
+  @Override
+  public Optional<FileLock> status(LockStatusCommandRequest request) {
+    String file = request.getFile();
+    try {
+      return getFileLock(file, open()).map(lock -> lock);
```
Comment (eheimbuch) [inline scm-plugins/scm-svn-plugin/src/main/java/sonia/scm/repository/spi/SvnFileLockCommand.java:112]: Remove this `map`

File: scm-plugins/scm-svn-plugin/src/test/java/sonia/scm/repository/spi/SvnFileLockCommandTest.java [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+  public void mockDefaultSubject() {
+    ThreadContext.bind(mockSubject("trillian"));
+  }
+
+  private Subject mockSubject(String name) {
+    Subject subject = mock(Subject.class);
+    ThreadContext.bind(subject);
+    when(subject.getPrincipal()).thenReturn(name);
+    return subject;
+  }
```
Comment (eheimbuch) [inline scm-plugins/scm-svn-plugin/src/test/java/sonia/scm/repository/spi/SvnFileLockCommandTest.java:60]: One `ThreadContext.bind(subject)` should be enough

File: scm-plugins/scm-svn-plugin/src/test/java/sonia/scm/repository/spi/SvnModifyCommandTest.java [hunk 4]
```diff
@@ -191,4 +194,54 @@ public void shouldPassIfExpectedRevisionMatchesCurrentRevision() throws IOExcept
...
+    initSecurityManager();
+
+    ModifyCommandRequest request = new ModifyCommandRequest();
+    request.addRequest(new ModifyCommandRequest.DeleteFileRequest("a.txt", false));
+    request.setCommitMessage("this should not happen");
+    request.setAuthor(new Person("Arthur Dent", "dent@hitchhiker.com"));
+
+    assertThrows(FileLockedException.class, () -> svnModifyCommand.execute(request));
+    WorkingCopy<File, File> workingCopy = workingCopyFactory.createWorkingCopy(context, null);
+    assertThat(new File(workingCopy.getWorkingRepository().getAbsolutePath() + "/a.txt")).exists();
```
Comment (eheimbuch) [inline scm-plugins/scm-svn-plugin/src/test/java/sonia/scm/repository/spi/SvnModifyCommandTest.java:215]: `ThreadContext.unbindSubject()`?

**Review by eheimbuch — CHANGES_REQUESTED** [review]

File: scm-plugins/scm-svn-plugin/src/main/java/sonia/scm/repository/spi/SvnFileLockCommand.java (hunk 2 above)
Comment (pfeuffer) [reply scm-plugins/scm-svn-plugin/src/main/java/sonia/scm/repository/spi/SvnFileLockCommand.java:112]: Believe this or not, this is necessary for the type check.
You can cast `SvnFileLock` to `FileLock`, which in this case is done implicitly by `map`. But you cannot cast `Optional<SvnFileLock>` to `Optional<FileLock>`.

File: scm-plugins/scm-svn-plugin/src/test/java/sonia/scm/repository/spi/SvnModifyCommandTest.java (hunk 4 above)
Comment (pfeuffer) [reply scm-plugins/scm-svn-plugin/src/test/java/sonia/scm/repository/spi/SvnModifyCommandTest.java:215]: This is done in `cleanUpSecurityManager` which is annotated with `@After`, or am I missing something?

Comment (eheimbuch) [reply scm-plugins/scm-svn-plugin/src/test/java/sonia/scm/repository/spi/SvnModifyCommandTest.java:215]: You are right, i missed this one in the diff view

**Pull request merged**